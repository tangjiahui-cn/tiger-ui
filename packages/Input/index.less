@import '../_style/theme';

.input {
  font-size: @fontSize;
  background-color: transparent;
  border: none;
  height: @fontSize;

  &::placeholder {
    color: @textHover;
  }

  &:focus {
    outline: none;
  }

  &-only {
    vertical-align: middle;
    transition: all @duration;
    border: 1px solid @borderColor;

    &:focus {
      border: 1px solid @primary;
    }
  }

  &-large {
    padding: @inputPaddingLarge;
    font-size: @fontSizeLarge;
    height: @fontSizeLarge;
  }

  &-middle {
    padding: @inputPadding;
  }

  &-small {
    padding: @inputPaddingSmall;
  }
}

.input-fix {
  &-prefix {
    margin-left: 12px;
  }

  &-suffix {
    margin-right: 12px;
  }

  & + .input {
    padding-left: 0;
    padding-right: 0;
    margin: 0 12px;
  }
}

.input-wrapper {
  position: relative;
  font-size: @fontSize;
  line-height: @fontSize;
  display: inline-block;
  transition: all @duration;
  vertical-align: middle;
  color: @color;

  &>.input-border {
    position: absolute;
    top: -1px;
    left: 0px;
    width: 100%;
    height: 100%;
    border: 1px solid @borderColor;
    pointer-events: none;
  }

  & > input:focus ~ .input-border {
    border: 1px solid @primary;
  }
}
