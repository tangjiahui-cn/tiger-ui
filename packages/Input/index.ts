/**
 * Input
 *
 * @author tangjiahui
 * @date 2024/4/19
 */
import Input from './input';

export default Input;
export type { InputProps, InputFC, InputSize, BaseInputProps, BaseInputPropsKeys } from './input';
