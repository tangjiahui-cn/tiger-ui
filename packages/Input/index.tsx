import React from "react";
import Styles from './index.less';

export interface InputProps {
  value: string;
}
export default function Input () {
  return <input className={Styles.my_input} placeholder={'请输入'}/>
}
