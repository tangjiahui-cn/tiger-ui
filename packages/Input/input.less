@import '../_theme/variable.less';

@prefix: prefix('input');
@prefixWrap: ~'@{prefix}-wrap';
@borderClass: ~'@{prefixWrap}-border';

@{prefix} {
  .size();
  .common();
  border: none;
  display: inline-block;

  &::placeholder {
    color: @placeholderColor;
  }

  &:focus {
    outline: none;
  }

  &:disabled {
    .disabled();
  }

  &-pure {
    transition: all @duration;
    border: 1px solid @borderColor;
    &:focus {
      border: 1px solid @primary;
    }
  }
}

@{prefixWrap} {
  .common();
  position: relative;
  display: inline-flex;
  transition: all @duration;
  align-items: center;

  & > @{borderClass} {
    position: absolute;
    top: -1px;
    left: -1px;
    width: 100%;
    height: 100%;
    border: 1px solid @borderColor;
    pointer-events: none;
    transition: border-color @duration;
  }

  & > input:focus ~ @{borderClass} {
    border: 1px solid @primary;
  }

  & input {
    flex: 1px;
    border: none;
  }

  &-disabled {
    .disabled();
  }

  &-prefix {
    .textEllipsis();
    margin-left: 12px;
  }

  &-suffix {
    .textEllipsis();
    margin-right: 12px;
  }
}

.disabled () {
  color: @colorDisabled;
  background-color: @colorBgDisabled;
  cursor: not-allowed;
}

.textEllipsis () {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.common () {
  width: 100%;
  color: @color;
  line-height: 1.125;
  font-size: @fontSize;
  box-sizing: border-box;
  vertical-align: middle;
  background-color: white;
}

.size () {
  &-large {
    padding: 12px;
    font-size: @fontSizeLarge;
  }

  &-middle {
    padding: 8px 12px;
  }

  &-small {
    padding: 4px 12px;
  }
}
