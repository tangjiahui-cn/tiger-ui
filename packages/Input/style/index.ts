import useEmotion from '../../_utils/hooks/useEmotion';
import usePrefix from '../../_utils/hooks/usePrefix';
import useToken from '../../_utils/hooks/useToken';
import brotherSelector from '../../_utils/style/brotherSelector';
import React from 'react';
import type { InputSize } from '..';
import childrenSelector from '@/_utils/style/childrenSelector';

type StyleObject = {
  [k: string]: any | StyleObject;
};

export function useStyle(componentName: string): {
  input: () => string;
  size: (size: InputSize) => string;
  pure: () => string;
  prefix: () => string;
  suffix: () => string;
  border: () => string;
  inputWrapper: () => string;
  wrapperDisabled: () => string;
} {
  const token = useToken();
  const prefix = usePrefix(componentName);
  const { css } = useEmotion({ key: prefix });
  const { css: cssWrapper } = useEmotion({ key: `${prefix}-wrapper` });
  const pure = () => `${prefix}-pure`;

  const textEllipsis = {
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
  };

  const disabled = {
    borderColor: token.colorDisabled,
    color: token.colorDisabled,
    backgroundColor: token.colorBgDisabled,
    cursor: 'not-allowed',
  };

  const commonInput: React.CSSProperties = {
    width: '100%',
    color: token.color,
    lineHeight: 1.125,
    fontSize: token.fontSize,
    boxSizing: 'border-box',
    verticalAlign: 'middle',
    backgroundColor: 'white',
  };

  function genSizeStyle() {
    const getClassName = (size: InputSize) => `${prefix}-${size}`;

    const sizeStyle = {
      [getClassName('large')]: {
        padding: 12,
        fontSize: token.fontSizeLarge,
      },
      [getClassName('middle')]: {
        padding: '8px 12px',
      },
      [getClassName('small')]: {
        padding: '4px 12px',
      },
    };

    return {
      size: getClassName,
      sizeStyle,
    };
  }

  const { size, sizeStyle } = genSizeStyle();
  const pureStyle = {
    [pure()]: {
      transition: `all ${token.duration}`,
      border: `1px solid ${token.borderColor}`,
      '&:focus': {
        border: `1px solid ${token.primary}`,
      },
    },
  };

  const inputPrefix = () => `${prefix}-prefix`;
  const inputPrefixStyle = {
    [inputPrefix()]: {
      ...textEllipsis,
      marginLeft: 12,
    },
  };

  const inputSuffix = () => `${prefix}-suffix`;
  const inputSuffixStyle = {
    [inputSuffix()]: {
      ...textEllipsis,
      marginRight: 12,
    },
  };

  const border = () => `${prefix}-border`;

  const wrapperDisabled = () => `${prefix}-wrapper-disabled`;
  const wrapperDisabledStyle = {
    [wrapperDisabled()]: disabled,
  };

  const input = () => {
    return css({
      ...commonInput,
      border: 'none',
      display: 'inline-block',
      '&::placeholder': {
        color: token.colorHover,
      },
      '&:focus': {
        outline: 'none',
      },
      '&:disabled': disabled,
      ...brotherSelector(sizeStyle),
      ...brotherSelector(pureStyle),
    });
  };

  const inputWrapper = () => {
    return cssWrapper({
      ...commonInput,
      position: 'relative',
      display: 'inline-flex',
      transition: `all ${token.duration}`,
      alignItems: 'center',
      border: '1px solid transparent',
      [`& > .${border()}`]: {
        position: 'absolute',
        top: -1,
        left: -1,
        width: '100%',
        height: '100%',
        border: `1px solid ${token.borderColor}`,
        pointerEvents: 'none',
      },
      [`& > input:focus ~ .${border()}`]: {
        border: `1px solid ${token.primary}`,
      },
      '& input': {
        flex: 1,
        border: 'none',
      },
      ...brotherSelector(wrapperDisabledStyle),
      ...childrenSelector(inputPrefixStyle),
      ...childrenSelector(inputSuffixStyle),
    });
  };

  return {
    input,
    size,
    pure,
    border,
    prefix: inputPrefix,
    suffix: inputSuffix,
    wrapperDisabled,
    inputWrapper,
  };
}
