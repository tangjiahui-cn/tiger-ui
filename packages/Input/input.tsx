/**
 * input
 *
 * @author tangjiahui
 * @date 2023/4/17
 */
import React, { DOMAttributes, RefAttributes } from 'react';
import { SizeType as InputSize } from '../_types/common';
import classNames from 'classnames';
import { omit } from '@/_utils/object';
import { usePrefix, useLocale } from '@/ConfigProvider/ConfigProvider';
import './input.less';

export type { InputSize };
export interface BaseInputProps {
  /** max input length */
  maxLength?: number;
  /** controlled value from outside */
  value?: string;
  /** input box size */
  size?: InputSize;
  /** input prefix */
  prefix?: React.ReactNode;
  /** input suffix */
  suffix?: React.ReactNode;
  /** input placeholder */
  placeholder?: string;
  /** if disabled */
  disabled?: boolean;
  /** prefix style */
  prefixStyle?: React.CSSProperties;
  /** suffix style */
  suffixStyle?: React.CSSProperties;
  /** change callback */
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  /** onKeydown event callback */
  onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement>) => void;
  /** onBlur event callback */
  onBlur?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  /** style */
  style?: React.CSSProperties;
  /** className */
  className?: string;
}

export type InputProps = BaseInputProps &
  React.HTMLAttributes<HTMLInputElement> &
  RefAttributes<HTMLInputElement>;

export type InputFC = React.ForwardRefExoticComponent<InputProps>;
const Input: InputFC = React.forwardRef(function (
  props: InputProps,
  ref: React.ForwardedRef<HTMLInputElement>,
) {
  const {
    maxLength,
    value,
    size,
    prefix,
    suffix,
    placeholder,
    disabled,
    prefixStyle,
    suffixStyle,
    onChange,
    onBlur,
    onKeyDown,
    style,
    className,
    ...rest
  } = props;
  const locale = useLocale();
  const prefixCls = usePrefix('input');
  const prefixWrapCls = `${prefixCls}-wrap`;

  const isPure = !(prefix || suffix);

  const InputEl = (
    <input
      {...(isPure ? rest : undefined)}
      ref={ref}
      maxLength={maxLength}
      style={(isPure && style) || undefined}
      disabled={disabled}
      className={classNames(
        isPure && className,
        prefixCls,
        `${prefixCls}-${size || 'middle'}`,
        isPure && `${prefixCls}-pure`,
      )}
      placeholder={placeholder || locale.inputPlaceholder}
      value={value}
      onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
        onChange?.(e);
      }}
      onBlur={(e: React.ChangeEvent<HTMLInputElement>) => {
        onBlur?.(e);
      }}
      onKeyDown={(e: React.KeyboardEvent<HTMLInputElement>) => {
        onKeyDown?.(e);
      }}
    />
  );

  if (isPure) {
    return InputEl;
  }

  return (
    <span
      {...(!isPure ? rest : undefined)}
      className={classNames(
        !isPure && className,
        prefixWrapCls,
        disabled && `${prefixWrapCls}-disabled`,
      )}
      style={style}
    >
      {prefix && (
        <span className={`${prefixWrapCls}-prefix`} style={prefixStyle}>
          {prefix}
        </span>
      )}
      {InputEl}
      {suffix && (
        <span className={`${prefixWrapCls}-suffix`} style={suffixStyle}>
          {suffix}
        </span>
      )}
      <span className={`${prefixWrapCls}-border`} />
    </span>
  );
});

export default Input;
