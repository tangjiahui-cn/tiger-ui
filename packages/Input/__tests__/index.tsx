import { cssSelectorTest, mountTest, renderTest } from '../../../tests';
import Input from '../index';
import { fireEvent, render } from '@testing-library/react';

describe('input', () => {
  renderTest(Input);
  mountTest(Input);
  mountTest(() => <Input value={'value'} />);
  mountTest(() => <Input size={'middle'} />);
  mountTest(() => <Input prefix={'prefix'} />);
  mountTest(() => <Input prefix={<div>prefix</div>} />);
  mountTest(() => <Input suffix={'suffix'} />);
  mountTest(() => <Input suffix={<div>suffix</div>} />);
  mountTest(() => <Input disabled />);
  mountTest(() => <Input placeholder={'请输入'} />);
  mountTest(() => (
    <Input
      value={'111'}
      size={'middle'}
      prefix={<div>prefix</div>}
      suffix={'suffix'}
      disabled
      placeholder={'请输入'}
    />
  ));

  describe('dom event test', () => {
    it('onchange trigger', () => {
      const mockFn = jest.fn();
      const wrapper = render(
        <Input
          onChange={(e) => {
            expect(e?.target?.value).toEqual('1111');
            mockFn();
          }}
        />,
      );
      const dom = wrapper.container.querySelector('input');
      expect(dom).toBeTruthy();
      if (dom) {
        // @ts-ignore
        fireEvent?.change?.(dom, {
          target: {
            value: '1111',
          },
        });
        expect(mockFn).toHaveBeenCalledTimes(1);
      }
    });
  });

  cssSelectorTest(['.input'], Input);
  cssSelectorTest(['.input-middle'], () => <Input size={'middle'} />);
  cssSelectorTest(['.input-fix-prefix'], () => <Input prefix={'xxx'} />);
  cssSelectorTest(['.input-fix-suffix'], () => <Input suffix={'yyy'} />);
  cssSelectorTest(['[value="111"]'], () => <Input value={'111'} />);
  cssSelectorTest(['[disabled]'], () => <Input disabled />);

  // test props 'prefix' and 'suffix'
  describe('props test', function () {
    it('prefix should <span>xxx</span>', () => {
      const wrapper = render(<Input prefix={'xxx'} />);
      const dom = wrapper.container.querySelector('.input-fix-prefix');
      expect(dom?.innerHTML).toEqual('xxx');
    });
    it('suffix should <span>yyy</span>', () => {
      const wrapper = render(<Input suffix={'yyy'} />);
      const dom = wrapper.container.querySelector('.input-fix-suffix');
      expect(dom?.innerHTML).toEqual('yyy');
    });
  });
});
