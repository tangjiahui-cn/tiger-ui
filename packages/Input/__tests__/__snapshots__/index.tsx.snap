// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input render test render correctly 1`] = `
<DocumentFragment>
  <input
    class="input input-middle input-pure"
    placeholder="请输入"
    value=""
  />
</DocumentFragment>
`;
